tests/e2e/cert_manager_test.go

```diff
@@ -0,0 +1,73 @@
+// Copyright © 2023 Cisco Systems, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	. "github.com/onsi/ginkgo/v2"
+)
+
+// requireApplyingCertManagerCRDs deploys the cert-manager CRDs and checks
+// their existence afterwards.
+func requireApplyingCertManagerCRDs(kubectlOptions *k8s.KubectlOptions, certManagerVersion Version) {
+	It("Applying cert-manager CRDs", func() {
+		By("Applying cert-manager CRDs with version " + certManagerVersion)
+		applyK8sResourceManifest(
+			kubectlOptions,
+			"https://github.com/jetstack/cert-manager/releases/download/"+certManagerVersion+"/cert-manager.crds.yaml",
+		)
+
+		By("Verifying cert-manager CRDs")
+		requireExistingCRDs(
+			kubectlOptions,
+			"certificaterequests.cert-manager.io",
+			"certificates.cert-manager.io",
+			"challenges.acme.cert-manager.io",
+			"clusterissuers.cert-manager.io",
+			"issuers.cert-manager.io",
+			"orders.acme.cert-manager.io",
+		)
+	})
+}
+
+// requireInstallingCertManager deploys cert-manager CRDs and Helm chart and
+// checks the success of those operations.
+func requireInstallingCertManager(kubectlOptions *k8s.KubectlOptions, certManagerVersion Version) {
+	When("Installing cert-manager", func() {
```

Comment:
In this case we need to use Ordered containers so the order of the "It"s will be fixed.
When("Installing cert-manager", Ordered, func() {

Reply:
I agree, though I'd also add `ContinueOnFailure` because of this:

> When a spec in an Ordered container fails all subsequent specs are skipped. Ginkgo will then run any AfterAll node closures to clean up after the specs.
> 
> You can override this behavior by decorating an Ordered container with ContinueOnFailure.

Reply:
I think we should use Fast fail at installation, kind cluster tests, kafkacluster deploy test, zookeeper cluster deploy test and at uninstallations (https://onsi.github.io/ginkgo/#mental-model-how-ginkgo-handles-failure)
Otherwise I agree to use the ContinueOnFailure in other cases.

Reply:
https://onsi.github.io/ginkgo/#spec-subjects-it
> Every Ginkgo spec has exactly one subject node. You can add a single spec to a suite by adding a new subject node using It(<description>, <closure>). Here's a spec to validate that we can extract the author's last name from a Book model:

 https://onsi.github.io/ginkgo/#organizing-specs-with-container-nodes
> Ginkgo allows you to hierarchically organize the specs in your suite using container nodes. Ginkgo provides three synonymous nouns for creating container nodes: Describe, Context, and When. These three are functionally identical and are provided to help the spec narrative flow. You usually Describe different capabilities of your code and explore the behavior of each capability across different Contexts.

https://onsi.github.io/ginkgo/#mental-model-how-ginkgo-traverses-the-spec-hierarchy
> When running a spec Ginkgo invokes the setup and subject nodes closures in the correct order and tracks any failed assertions.

https://onsi.github.io/ginkgo/#spec-randomization
> Ginkgo's default behavior is to only randomize the order of top-level containers -- the specs within those containers continue to run in the order in which they are specified in the test files. This is helpful when developing specs as it mitigates the cognitive overload of having specs within a container continuously change the order in which they run during a debugging session.

By default Ginkgo behavior the order of the Its will always be fixed according to the Ginkgo docs - because they are never top level containers, they are subject nodes under at least one container node. Only Whens/Contexts/Describes are parallelized.
`It` is considered a subject node while `When` is considered a container node.
See the example in my 3rd quote, I think we are analogous to that.

That's why I designed the solution so the highest level test specifications such as "external producer-consumer" test are going to be the top level containers which are going to be randomized.
1. which eliminates the need to specify ordered in lower level whens (that would be redundant/noisy, but we can opt for it if we want to) and also
2. which is supposed to be fine because with slow-fail we don't care if the failure order is InstallFailed SimpleKafkaClusterDeployFailed ExternalProducerConsumerFailed or the other way around, all will be run and logged anyway and we need to check which is the first to fix.

We can still opt for ordered run of these if we want to make sure to have these in an orderly fashion in the log, but IMO these should be run in parallel anyway on different clusters - to quicken up the execution - which will cause the order to be somewhat messed up anyway, but slowfail will help us be able to determine the smallest test specification in featureset that failed to start investigating upon.

(I do agree if we introduce the order then we need to introduce the `ContinueOnFailure` as well.)

Reply:
I also checked and I missed something. We dont need to use ordered container in this case.
We need to use only here (https://github.com/banzaicloud/koperator/blob/f7fe06f20be0d303c28575169aaaa0ef1f0d1888/tests/e2e/install_test.go#L23) because to level containers here will be randomized and uninstallation will happen sooner then installation :)

Reply:
Hmmm, in that case I'm the one who misunderstood something - I thought `It`s should always be ordered inside a `When` and only top level `When`s should be randomized, but I believe your example on Koperator - and then the same should hold true for cert-manager (CRD and Helm chart) and then I would say we should do the same for all (prometheus and zookeeper included).
I will get this fixed.

Reply:
Fixed in 64f5aac.

TBH I don't fully understand how Ginkgo work and don't have the time to figure out so I added the Ordered decorator to each `When` to be sure, that's a safer pattern to follow for now, if we find a better solution later we can easily search & replace.